server/src/main/java/org/opensearch/indices/replication/checkpoint/PublishCheckpointAction.java

```diff
@@ -118,7 +118,12 @@ final void publish(IndexShard indexShard) {
             PublishCheckpointRequest request = new PublishCheckpointRequest(indexShard.getLatestReplicationCheckpoint());
             final ReplicationCheckpoint checkpoint = request.getCheckpoint();
 
-            final List<ShardRouting> replicationTargets = indexShard.getReplicationGroup().getReplicationTargets();
+            final List<ShardRouting> replicationTargets;
+            try {
+                replicationTargets = indexShard.getReplicationGroup().getReplicationTargets();
```

Comment:
@Rishikesh1159 : Thanks for fixing this. Can you also please add more details around why this fix, reduces/fixes the flakyness here.

Reply:
Sure. During the process of relocation of primary, primary still publishes checkpoints on refresh even if relocation is in progress. So, during relocation when handoff is completed we switch the primarymode to false. So, this handoff process and publishing checkpoints to replicas happen parallely in our scenario.

Before publishing checkpoints to replicas we do call getReplicationGroup() on IndexShard. And this methods asserts on primarymode. So, very rarely we end up in a situation where primarymode of shard is switched between checkpointRefreshListeners afterRefresh check [here](https://github.com/opensearch-project/OpenSearch/blob/b0ab550ecfb80a2942b58169c5c38e500950a8cc/server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java#L43) and this getReplicationGroup() assert. 

By throwing an exception in getReplicationGroup() when shard not in primary mode we are reducing flakiness. But again we immediately have assert after this primarymode check in getReplicationGroup(), If relocation handoff completes between these two lines and primarymode switches, we may again fail assert, so there is flakiness but will not happen often.

We are not concerned about primary sending checkpoints to replica here, we are only trying to reduce flakiness.